pkg/reader/usdc_reader.go

```diff
@@ -338,6 +360,148 @@ func (u evmUSDCMessageReader) recreateMessageTransmitterEvents(
 	return messageTransmitterEvents, nil
 }
 
+type solanaUSDCMessageReader struct {
+	lggr           logger.Logger
+	contractReader contractreader.Extended
+	cctpDestDomain map[uint64]uint32
+}
+
+// getMessageTokenData extracts token data from the CCTP MessageSent event.
+func (u solanaUSDCMessageReader) getMessageTokenData(
+	tokens map[MessageTokenID]cciptypes.RampTokenAmount,
+) (map[MessageTokenID]SourceTokenDataPayload, error) {
+	messageTransmitterEvents := make(map[MessageTokenID]SourceTokenDataPayload)
+
+	for id, token := range tokens {
+		sourceTokenPayload, err := extractABIPayload(token.ExtraData)
+		if err != nil {
+			return nil, err
+		}
+		messageTransmitterEvents[id] = *sourceTokenPayload
+	}
+	return messageTransmitterEvents, nil
+}
+
+// SolanaCCTPUSDCMessageEvent matches the onChain CcipCctpMessageSentEvent.
+type SolanaCCTPUSDCMessageEvent struct {
+	Discriminator       [8]byte
+	OriginalSender      [32]byte
+	RemoteChainSelector uint64
+	MsgTotalNonce       uint64
+	EventAddress        [32]byte
+	SourceDomain        uint32
+	CctpNonce           uint64
+	MessageSentBytes    []byte
+}
+
+// MessagesByTokenID retrieves the CCTP MessageSent events.
+func (u solanaUSDCMessageReader) MessagesByTokenID(
+	ctx context.Context,
+	source, dest cciptypes.ChainSelector,
+	tokens map[MessageTokenID]cciptypes.RampTokenAmount,
+) (map[MessageTokenID]cciptypes.Bytes, error) {
+	if len(tokens) == 0 {
+		return map[MessageTokenID]cciptypes.Bytes{}, nil
+	}
+	u.lggr.Debugw("Searching for Solana CCTP USDC logs",
+		"numExpected", len(tokens))
+
+	// Parse the extra data field to get the CCTP nonces and source domains.
+	cctpData, err := u.getMessageTokenData(tokens)
+	if err != nil {
+		return nil, err
+	}
+
+	// Query the token pool contract for the MessageSent event data.
+	expressions := []query.Expression{query.Confidence(primitives.Finalized)}
+	for _, data := range cctpData {
+		// This is much more expensive than the EVM version. Rather than a
+		// single ANY expression, we have separate expressions for each
+		// nonce and source domain pair. This is because Solana doesn't have
+		// a combined ID like EVM does.
+		// TODO: optimize. CR modifier or a new field in our event.
+		expressions = append(expressions, query.And(
+			query.Comparator(
+				consts.EventAttributeCCTPNonce,
+				primitives.ValueComparator{
+					Value:    data.Nonce,
+					Operator: primitives.Eq,
+				},
+			),
+			query.Comparator(
+				consts.EventAttributeSourceDomain,
+				primitives.ValueComparator{
+					Value:    data.SourceDomain,
+					Operator: primitives.Eq,
+				},
+			),
+		))
+	}
+
+	// Parent expressions for the query.
+	keyFilter, err := query.Where(
+		// Using same as EVM for consistency. Only used by off-chain components so name does not have to align with on-chain
+		consts.EventNameCCTPMessageSent,
+		expressions...,
```

Comment:
I think this logic could break if we get multiple CCTP data. The default operator in the LogPoller is `AND` so stringing expressions together like this would make a query like `nonce=1 && domain=5 && nonce=2 && domain=5`. I think we probably need to use the Or comparator here
```suggestion
		query.Or(expressions...),
```

Reply:
Oh and we should probably move the `query.Confidence(primitives.Finalized)` from above here so it doesn't get mixed up the with the nonce/domain queries
```suggestion
		query.Confidence(primitives.Finalized),
		expressions...,
```

Reply:
Unresolving this since I noticed the `query.Confidence(primitives.Finalized)` expression wasn't moved down into the where clause